rfcs/031-relation_batcher/README.md

```diff
@@ -0,0 +1,112 @@
+
+### Router
+
+The `router` takes a work id, looks it up in the `pipeline_storage` and checks if it's part of an 
+archive (i.e. has a `collectionPath` field).  If it is, it sends the value of `collectionPath` to the `batcher`, otherwise sending the work straight to the `id_minter_works`. 
+
+### Batcher
+
+The `batcher` groups work received over a given time period and works out the optimal set of paths to send towards 
+the `relation_embedder`. It is deployed as a single task and runs at intervals of fixed duration (say, 30  minutes), reading all messages from the queue in a single pass. 
+
+Assuming a graph shape like this:
+
+```
+A
+|-----
+|    |
+B    C
+|    |-----
+|    |    |
+D    E    F
+|
+G
+|
+H
+```
+
+With a queue containing the paths for works `H`, `D`, `E` and `F`, the `batcher` would:
+
+- Determine the parent set of each of the messages in the queue (ie. in the example `G`, `B`, `C`)
+- Filter out nodes which are descendants from other ones (i,e, `G` as it is a descendent of `B`) 
+- Send the remaining to the `relation_embedder` (`B` and `C` )
+
+
+
+In the example above of this archive tree:
+
+```
+MS9225
+|
+|---
+|  |
+1  2
+   |-------------
+   |  |  |      |
+   1  2  3 ... 3583
+```
+
+During a reindex in the best case scenario all works in this archive are read from queue at the same time in the `batcher`, and the root work (`MS9225`) is sent once to the `relation_embedder`. In a more plausible scenario where the works are split across multiple batches, each batch will contain the tree only once, which although it will lead to duplicate work this will still drastically reduce the amount of work performed.
+
+### Relation embedder
+
+The `relation_embedder` receives messages each containing a `collectionPath` as input from the `batcher`, instead of work IDs as previously. It is unchanged in every other aspect.
+
+## Potential Issues
+
+* The `batcher` necessarily introduces some latency into the processing of archive works that is not present in the current pipeline. We are able to tweak the time processing interval for this to provide a compromise between `relation_embedder` performance, delay, and costs. One potential solution would be to make the interval duration change dynamically in response to pipeline load, so that it's big while there are many messages  in flight, but it's small if there are not many messages, decreasing latency for updates while retaining efficiency during a reindex.
+
+* Another disadvantage is the proposal introduces 2 new services, and some additional architectural complexity due to the fact that the messages split at the `router` before joining again at the `id_minter`. The exact number of services and modes of communication could probably be modified, but the particular design we came up with was the simplest we could think of in regards to not introducing complicated logic and additional storage beyond using queues with SQS.
```

Comment:
Copying relevant parts of discussion from slack here:

Reply:
I think that this RFC proposes an implicit split between the tree-building and the denormalisation stage; the batcher builds a tree in memory in which only parent->child references require updates in order to avoid our need to denormalise entire lineages on every update.

I would be interested in exploring solutions which make this 2-step process explicit, hopefully avoiding complexity in the process.

Reply:
> @kenoir I think we need to talk about this, it feels like we might need to do something different with relationships to get this to work rather than trying to shoehorn it into the existing pipeline
> @jamesgorrie A lot of the work around concepts in the pipeline pointed in that direction too.
> @jamieparkinson It feels like tree building should be completed before denormalisation happens - as all you need to represent a tree is to contain child references in parents. I think this is really what the proposed batcher is doing (just doing so implicitly and - intentionally - suboptimally)
> @kenoir so something like reindex entirely without denormalisation then iterate over all works denormalising data onto them as appropriate?
> @jamieparkinson something like that - maybe with some intermediate storage, but then again I can’t actually think of why that would be necessary… trees are really simple data structures, we don’t need anything fancy to represent them
> @jamieparkinson maybe we end up with 2 indices, one denormalised and one not
> @kenoir run the denormalisation in process on a schedule for all works (or for just those that have changed in a given time period)
> @jamesgorrie this is where we got to with concepts. Trying to imagine the pipeline as: `source data (calm etc) => transforms => works index` and then `works index => transforms => denormalised works`

Reply:
To address the tree building issue: as me and @alicefuzier discussed on a call with @kenoir, building the trees is really not the issue. Having a seperate graph data store or what have you would not change the nature of the work, we already have a representation of trees that is queryable (i.e. as a `collectionPath`), and having a tree built before denormalisation would still require the same amount of work.

Reply:
> so something like reindex entirely without denormalisation then iterate over all works denormalising data onto them as appropriate?

> run the denormalisation in process on a schedule for all works (or for just those that have changed in a given time period)

I would argue that this is kind of what the the proposal in this RFC is doing. If we only wanted the denormalisation process to happen daily or whatever we could set the interval to 24 hours.

> Trying to imagine the pipeline as: source data (calm etc) => transforms => works index and then works index => transforms => denormalised works

I also think this is what the pipeline currently does. For me having an explicit split between these two parts of the pipeline does not really give anything: maybe you could deploy them separately maybe, but not sure whether that would be helpful or a hindrance.

Reply:
> I think that this RFC proposes an implicit split between the tree-building and the denormalisation stage; the batcher builds a tree in memory in which only parent->child references require updates in order to avoid our need to denormalise entire lineages on every update.
> I would be interested in exploring solutions which make this 2-step process explicit, hopefully avoiding complexity in the process.

Making this more explicit maybe be useful in future (as I discussed in the original [Relation Embedder RFC](https://github.com/wellcomecollection/docs/tree/master/rfcs/027-relation-embedder#graph-store)).

However imo it does not avoid any complexity and would definitely add some: having a extra store introduces more state into the pipeline, requires more infra, code for interacting with new databases etc.

Reply:
> I think trying to catch works that are related by slicing loading/processing messages into half an hour chunks, we might be better off running over the entire set of works after they've been reindexed and embedding relations then (after we've built the intermediate works store).

What I think this proposes is modifying the existing index (i.e. the one the API is reading from) at regular intervals. @kenoir can you explain why you think this would be better? To me this has lots of downsides:

* No longer a uni directional flow of data from inputs to outputs
* Separate to the main pipeline
* Puts more strain on the index that we are serving the API from

Reply:
Sorry for so many seperate messages. I am not being critical of all the above due to being wedded to the current design, just need to be convinced an alternative is an improvement / more complexity is worth it.

Reply:
>we do this on an intermediate store. 

I do mean on an intermediate store, yes.

Reply:
> having a tree built before denormalisation would still require the same amount of work.

The same amount of work as the solution proposed in this RFC? I think it would be slightly less due to the fact that there would be guaranteed no duplicated effort, right?

Regardless I'm not sure what you mean by "the tree building issue" - as you say this part is easy, that's why I was suggesting separating the denormalisation (the hard part) from it entirely.

Reply:
Maybe we are talking about different things here. Building a tree up front does not affect the number of opera the relation embedder has no to perform: you still query for other affected works (i.e. the subtree from its parent downwards), and then for every one of those works you query for its relations.  It just affects the way you make those two queries (whether using paths or another graph representation). The performance issue currently is not the query times but the fact that works are denormoralised multiple times, doing that upfront does not affect it.

Reply:
> I was suggesting separating the denormalisation (the hard part) from it entirely.

I don't think I follow what you mean. Is it something like @kenoir suggestion to do denormalisation only after having build an intermediate storage of all works without denormalisation?

Reply:
What I'm saying is effectively the same as @kenoir's suggestion, yes.

I am saying that building a tree up front means that you have enough information afterwards to perform an optimal denormalisation whereas denormalising as you go causes duplicate work as the tree is built - is that wrong?

Reply:
There is no concept of up front though really, as works keeping coming in. So the best you can do is within a time interval (although maybe can get more clever with respect to reindexing etc)

Reply:
In a reindex, with no dynamic updates coming through, the optimal would be to send the root of each archive after its descendants have all gone through. So if you just stopped at an intermediate store, you could just do that and you would have the relation_embedder building each tree exactly once (correct me if I'm wrong @warrd )

But we have dynamic updates coming from all sources which could potentially affect any denormalisation, so as @warrd is saying, it's a bit harder in that scenario to figure out when/how to stop cause it's a constant stream of updates.

Even for the reindex scenario, I think stopping at an intermediate stage only adds delay. It is what the batcher would be doing anyway and if you made the timeframe long enough you would hit the best case scenario. But then you would make the time that it takes to have a usable index bigger I think

Reply:
Yeah, I think I see now that separating the denormalisation step doesn't really offer anything better than an in-pipeline batching solution. Still want to discuss further whether a non-scalable batcher working on a time period is the best way to do that though.